Accurately set the lft and rght fields for moveDown

AD7six opened the pull request:
Needs similar updates to moveUp, fixes https://github.com/cakephp/cakephp/issues/5863.

Will cause some test fails because.. it looks like [this method didn't say what it was doing - but instead reported something different](https://github.com/cakephp/cakephp/compare/3.0...3.0-tree-wrong).

AD7six committed:
Accurately set the lft and rght fields as you go

Needs similar updates to moveUp

AD7six committed:
Update moveUp too

markstory commented:
Seems reasonable to me.

ADmad commented:
Tests failing.

AD7six commented:
> Tests failing.

But not the added ones :trollface: 

I need to pick the test case apart to determine which fails need fixing, and which passes are actually broken (if any).

ADmad commented:
For one failing test the actual value for lft is negative so something is fishy :stuck_out_tongue:

AD7six committed:
Actually, that can be a lot simpler

AD7six committed:
Remove assertTreeNumbers

E.g. It wouldn't indicate any difference between:

    1
    2
    3
    4

and

    1
    └──2
       ├──3
       └──4

Replace with a more descriptive and grokable custom assert method

AD7six committed:
Adjust all tests to use assertMpttValues

In addition to any other asserts that are performed

AD7six committed:
Single-step moveDown

AD7six committed:
Single-step moveUp

AD7six committed:
Use some better named variables to aide readability

AD7six committed:
Better node-to-hole logic

More readable at least

AD7six committed:
Prevent negative offsets

Not sure why is this is considered "valid" - there are tests expecting a
node returned, rather than false

AD7six commented:
The original problem this PR fixed was that moveDown - would return a node with lft/rght set to different values than were on the database, and the test case would only be asserting the nodes (transient) state whilst leaving the db in a _different_ state (saying it moved a node 2 places, and instead moving it 1).

This PR has morphed a bit to include:

 * Updating the test case to be easier to read/understand/maintain.
 * Update moveUp and moveDown to be easier to read/understand/maintain.
 * Updating moveUp and moveDown to move in one step (3 update queries) instead of once per move ( 3 * $number queries)

@lorenzo Because I've changed the implementation a fair bit - can you give it a look over please.

AD7six commented on src/ORM/Behavior/TreeBehavior.php:
```diff
@@ -557,51 +557,58 @@ public function moveUp(Entity $node, $number = 1)
-        $parentLeft = 0;
-        if ($node->get($parent)) {
-            $parentLeft = $this->_getNode($node->get($parent))->get($left);
+        if ($number < 0) {
```
@lorenzo this is one of the things I wanted specifically to check - it still returns the node, [as it did before](https://github.com/cakephp/cakephp/blob/3.0/tests/TestCase/ORM/Behavior/TreeBehaviorTest.php#L290) - is that desired behavior?

markstory replied:
I think the intent would have been to catch negatives as well. Negative values make no sense in MPTT.

AD7six replied:
I agree, though the same applies whatever the implementation (move this foo up -1 places).

I'll change it to treat < 1 as invalid.

markstory commented on tests/TestCase/ORM/Behavior/TreeBehaviorTest.php:
```diff
@@ -284,17 +339,49 @@ public function testMoveUp()
+            '___ 6: 7 -  5:Link 5',
+            '11:14 -  6:Link 6',
+            '_12:13 -  7:Link 7',
+            '15:16 -  8:Link 8'
```
I really like this style for the tests. Much simpler to read.

AD7six committed:
Don't accept negative moves

And bail early, rather than deal with it later

AD7six commented:
I think this is good enough to merge, if it breaks something it should hopefully be easier now to write a test for whatever scenario and fix it.
